src/shared/elf-util.c

```diff
@@ -440,6 +440,131 @@ static int parse_core(int fd, const char *executable, char **ret, JsonVariant **
+                        goto finish;
+                }
+        }
+
+        elf_version(EV_CURRENT);
+
+        c.elf = elf_begin(fd, ELF_C_READ_MMAP, NULL);
+        if (!c.elf) {
+                r = -EINVAL;
+                goto finish;
+        }
+
+        if (!gelf_getehdr(c.elf, &elf_header)) {
+                r = -EINVAL;
+                goto finish;
+        }
+
+        /* Note that e_type is always DYN for both executables and libraries, so we can't tell them apart from the header. */
+        r = json_build(&elf_metadata,
+                       JSON_BUILD_OBJECT(JSON_BUILD_PAIR("elfType", JSON_BUILD_STRING(elf_header.e_type == ET_CORE ? "coredump" : "binary"))));
+        if (r < 0)
+                goto finish;
+        if (ret)
+                fprintf(c.f, "ELF object type: %s\n", elf_header.e_type == ET_CORE ? "coredump" : "binary");
+
+        elf_architecture = dwelf_elf_e_machine_string(elf_header.e_machine);
+        if (elf_architecture) {
+                _cleanup_(json_variant_unrefp) JsonVariant *json_architecture = NULL;
+
+                r = json_build(&json_architecture,
+                                JSON_BUILD_OBJECT(JSON_BUILD_PAIR("elfArchitecture", JSON_BUILD_STRING(elf_architecture))));
+                if (r < 0)
+                        goto finish;
+
+                r = json_variant_merge(&elf_metadata, json_architecture);
+                if (r < 0)
+                        goto finish;
+
+                if (ret)
+                        fprintf(c.f, "ELF object binary architecture: %s\n", elf_architecture);
+        }
+
+        if (elf_header.e_type == ET_CORE) {
+                _cleanup_free_ char *out = NULL;
+
+                r = parse_core(fd, executable, ret ? &out : NULL, &package_metadata);
+                if (r < 0)
+                        goto finish;
+
+                if (out)
+                        fprintf(c.f, "%s", out);
+        } else {
+                _cleanup_(json_variant_unrefp) JsonVariant *id_json = NULL;
+
+                r = parse_buildid(NULL, c.elf, executable, &c, &id_json);
+                if (r < 0)
+                        goto finish;
+
+                r = parse_package_metadata(executable, id_json, c.elf, &c);
+                if (r < 0)
+                        goto finish;
+
+                /* If we found a build-id and nothing else, return at least that. */
+                if (!package_metadata && id_json) {
+                        r = json_build(&package_metadata, JSON_BUILD_OBJECT(JSON_BUILD_PAIR(executable, JSON_BUILD_VARIANT(id_json))));
+                        if (r < 0)
+                                goto finish;
+                }
+        }
+
+        /* We always at least have the ELF type, so merge that (and possibly the arch). */
+        r = json_variant_merge(&elf_metadata, package_metadata);
+        if (r < 0)
+                goto finish;
+
+        if (ret)
+                *ret = TAKE_PTR(buf);
```

Comment:
`buf` is not valid here yet.

```diff
diff --git src/shared/elf-util.c src/shared/elf-util.c
index f5af2968f0..59f5c67404 100644
--- src/shared/elf-util.c
+++ src/shared/elf-util.c
@@ -555,14 +555,14 @@ finish:
 static int parse_elf(int fd, const char *executable, char **ret, JsonVariant **ret_package_metadata) {
         _cleanup_(json_variant_unrefp) JsonVariant *package_metadata = NULL, *elf_metadata = NULL;
         _cleanup_(set_freep) Set *modules = NULL;
+        _cleanup_free_ char *buf = NULL;
+        size_t sz = 0;
         struct stack_context c = {
                 .package_metadata = &package_metadata,
                 .modules = &modules,
         };
         const char *elf_architecture;
         GElf_Ehdr elf_header;
-        char *buf = NULL;
-        size_t sz = 0;
         int r;
 
         assert(fd >= 0);
@@ -651,6 +651,8 @@ static int parse_elf(int fd, const char *executable, char **ret, JsonVariant **r
         if (r < 0)
                 goto finish;

+        r = fflush_and_check(c.f);
+        if (r < 0)
+                goto finish;
+        c.f = safe_fclose(c.f);
+
         if (ret)
                 *ret = TAKE_PTR(buf);
         if (ret_package_metadata)
@@ -661,14 +663,10 @@ static int parse_elf(int fd, const char *executable, char **ret, JsonVariant **r
 finish:
         if (c.dwfl)
                 sym_dwfl_end(c.dwfl);
-
         if (c.elf)
                 sym_elf_end(c.elf);
-
         safe_fclose(c.f);
 
-        free(buf);
-
         if (r == -EINVAL)
                 log_warning("Failed to inspect ELF: %s", sym_dwfl_errmsg(sym_dwfl_errno()));
         else if (r < 0)
```

Reply:
good catch - will fix with a separate commit, so we can backport it

Reply:
I thought that the bug was introduced in one of the patches in this PR (the one that adds this code).

Reply:
it was copied from the code that handles the core: https://github.com/systemd/systemd/pull/21454/commits/9bc785094a1b897179ccaa76e4fd43c9d23c3b90

Reply:
Oh, OK. So there are two fixes here: one is that `buf` may be invalidated when the file is closed. This is important, and the old code didn't show the issue. The second one is the addition of `fflush_and_check()`. This is more of a theoretical issue: apart from oom conditions, it should never matter.